Bronze age

AndrewVieira opened the pull request:
Hi there are a few things that need work still, but I wanted someone to code review my work so far.

WHAT'S NEW
1. When smithing, you have three choices that pop up in a modal window. The first green box is for bronze dagger, 2nd is the bronze sword, third is the bronze scimitar. Press the small white box to close without choosing. This shapes are just stand ins for now. I will use real images later.
2. Right click anvil to change your choice with "Choose" option
3. A new modal class for choosing recipes (like smithing)

WHAT NEEDS REVIEW
1. Any code that can be improved
2. Any bugs you find
4. What code should stay in modal-window,.js and what code should go to future child class smithing-interface.js
5. How do I use images for my modal class? (I tried using sprites but nothing showed up)
6. Anything else needed

KNOWN BUGS
Hammer hand animation works when smithing dagger, but not sword or scimitar

CurtisGreen commented on src/ui/modal-window.js:
```diff
@@ -0,0 +1,96 @@
+
+    addElements(elements) {
+        let offset = 8;
+        this.elementEvents = elements.map((obj) => obj);
```
What's the purpose of this, to copy them? If so `this.elementEvents = [...elements];` is functionally equivalent

CurtisGreen commented on src/ui/modals/smithing-interface.js:
```diff
@@ -0,0 +1,97 @@
+/* import { ModalWindow } from "../modal-window.js";
```
I think we should delete the file instead of just commenting stuff out, can look back at git history if we need the code again

CurtisGreen commented on src/ui/modal-window.js:
```diff
@@ -0,0 +1,96 @@
+            
+            let elementButton = new Button(this.scene, this.x + offset, this.y + 16, 48, 48);
+            elementButton.on("pointerup", () => {
+                if (this.visible == true) {
```
Nitpick
```suggestion
                if (this.visible) {
```

CurtisGreen commented on src/ui/modal-window.js:
```diff
@@ -0,0 +1,96 @@
+        let offset = 8;
+        this.elementEvents = elements.map((obj) => obj);
+
+        for (let i = 0; i < elements.length; i++) {
```
If you don't need the index you can do this:
```suggestion
        for (let element of elements) {
```

CurtisGreen replied:
Same for other indexed loops

CurtisGreen reviewed (changes requested):
Code pretty good so far just a few minor style changes. I haven't tested it yet, will try to do that tomorrow 👍

CurtisGreen commented on src/targets/anvil.js:
```diff
@@ -27,46 +31,85 @@ export class Anvil extends ClickableObject {
         this.sprite = new Button(scene, x, y, width, height);
         this.sprite.on("pointerdown", (pointer) => {
-            if (pointer.rightButtonDown() && !pointer.leftButtonDown()) {
+            if (pointer.rightButtonDown() && !pointer.leftButtonDown() && this.modalWindow.visible == false) {
```
```suggestion
            if (pointer.rightButtonDown() && !pointer.leftButtonDown() && !this.modalWindow.visible) {
```

CurtisGreen commented on src/ui/modal-window.js:
```diff
@@ -0,0 +1,96 @@
+            elementIcon.fillRect(this.x + offset, this.y + 16, 48, 48);
+            elementIcon.visible = this.visible;
+
+            offset += 56;
```
Would be better if offset was a constant based off of the loop index rather than adding to it each iteration. Makes it more obvious what its value is

AndrewVieira commented:
Thanks for reviewing the code. I will start making changes tomorrow.

EthanVieira commented on src/items/equipment/scimitars/bronze-scimitar.js:
```diff
@@ -1,6 +1,11 @@
 import Scimitar from "./scimitar.js";
 
 export default class BronzeScimitar extends Scimitar {
+    xp = 30.0;
```
Gives 25xp in RS and requires 2 bars

AndrewVieira replied:
I think I got my data from an incorrect source. Could you tell me what source you are using?

EthanVieira replied:
Use the OSRS wiki instead of the main RS wiki

EthanVieira commented on src/items/equipment/swords/bronze-sword.js:
```diff
@@ -1,6 +1,11 @@
 import Sword from "./sword.js";
 
 export default class BronzeSword extends Sword {
+    xp = 30.0;
```
12.5xp

EthanVieira commented on src/targets/anvil.js:
```diff
@@ -27,46 +31,85 @@ export class Anvil extends ClickableObject {
     }
 
     async clickTarget() {
+        if (this.modalWindow.visible == true) {
```
if (this.modalWindow.visible)

EthanVieira commented on src/targets/anvil.js:
```diff
@@ -27,46 +31,85 @@ export class Anvil extends ClickableObject {
+    }
+
+    async selectRecipe(itemName) {
+        if (this.hasMaterials() == false) {
```
if (!this.hasMaterials())

EthanVieira commented on src/targets/anvil.js:
```diff
@@ -27,46 +31,85 @@ export class Anvil extends ClickableObject {
 
     // Take bar and turn it into a smithable item
     async smith(itemName) {
+        if (this.hasMaterials() == false) {
```
if (!this.hasMaterials())

EthanVieira commented on src/targets/anvil.js:
```diff
@@ -27,46 +31,85 @@ export class Anvil extends ClickableObject {
         switch (selectedItem.name) {
             case "Bronze Bar":
-                this.smith("BronzeDagger");
+                    elements = ["BronzeDagger", "BronzeSword", "BronzeScimitar"];
```
I think I'm still in favor of having all of the modal window specifics inside of a child modal window class. i.e. bronze-smithing-modal-window.js or if all of the different smithing modals will be similar then maybe a smithing-modal-window that takes "bronze" as an argument. 

I prefer it that way for two reasons:
1) the modal window will be used for other things like the shop which, in addition to having different elements, could have a different background, multiple tabs, whether or not it has a scrollbar etc. Having to insert all that configurable info every time we create a modal window would be messy as opposed to just a single statement creating a child modal window class that has all the info inside.
2) the OOP model is so deeply ingrained in everything else we are doing it might as well continue

@CurtisGreen thoughts?

AndrewVieira replied:
I think your approach is best. I was overengineering a bit. Some future modal windows like shop and bank need to be a little customizable since the options can vary.

CurtisGreen replied:
Hard to say until we make a new modal and see what we need/don't need to be generic

EthanVieira reviewed (changes requested):
Naming convention for Bronze spear and warhammer icons should follow the rest. 

Don't need "Detail" on BronzeFullHelmDetail.png".

BronzeArrowtips -> capitalize Tips (in models)

Something I want to start doing is having Data files that can be loaded in place of default-data.js so we can jump to certain parts of the game to easily test them. In the next pr, could you create a test-smithing-data.js (or something similar)? Basically just copy default-data.js and then add in the additional things that would be useful for testing smithing. It would have all the smithing locations unlocked and plenty of bars/ore in the inventory to start with.

Overall great job! I look forward to using this!

CurtisGreen reviewed (changes requested):
Finally got around to testing, sorry for being slow.

I see an error in the console logs here whenever I click on the anvil without having anything selected first.
https://i.imgur.com/6ITAMnJ.png 

Same error for the furnace
https://i.imgur.com/DdvQ89L.png

Error when clicking hammer then clicking anvil
https://i.imgur.com/EN096ab.png

Modal images didn't load properly. Maybe you didn't push them up?
https://i.imgur.com/t0EfRqv.png

AndrewVieira commented:
I fixed all the issues you guys mentioned and I fixed the animation bug too. I will be working on the modal class and adding images.

CurtisGreen commented:
> I fixed all the issues you guys mentioned and I fixed the animation bug too. I will be working on the modal class and adding images.

Sweet, when you get the chance can you resolve the comments you've fixed so we can know what has/hasn't been done

AndrewVieira commented:
PROGRESS PICTURE: https://imgur.com/a/gnmf1W5
I'm getting pretty close to being done.

WHAT'S NEW
1. Added images and cleaned up the interface (icon position can be moved around arbitrarily in code) 
2. Item is not smithable if not high enough smithing level or enough bronze bars. Item text changes black if not high enough level. Bar count text turns orange if not enough bars.
3. Adding smithing test data file to speed up testing

TO DO
1. Add remaining bronze smithing items and organize them.
2. Add comments in log if not enough bars or high enough level? (Do you want me to add this?)
3. Any other issues you guys find

Please let me know what needs fixing.

CurtisGreen commented on src/cookie-io.js:
```diff
@@ -1,10 +1,12 @@
-import { defaultData } from "./default-data.js";
+// import { defaultData } from "./default-data.js";
+import { testSmithingData } from "./test-smithing-data.js"
```
You left some testing data in the commit

CurtisGreen commented on src/cookie-io.js:
```diff
@@ -1,10 +1,12 @@
     // Reset data (deep copy)
-    return JSON.parse(JSON.stringify(defaultData));
+    //return JSON.parse(JSON.stringify(defaultData));
+    return JSON.parse(JSON.stringify(testSmithingData));
```
here too

CurtisGreen commented on src/scenes/level.js:
```diff
@@ -155,6 +155,11 @@ export class LevelScene extends Phaser.Scene {
 
     enemyKilled(name) {
+        // Check if quest even exists
+        if (this.questAmounts[name] === undefined) {
```
1) This can be simplified to `if(!this.questAmounts[name])` since undefined, null, and false all resolve to false in an if statement
2) If something is breaking by calling this function we should fix it rather than working around the bug. I'm fine with keeping it but add a console log under it describing the error like cookie-io does

AndrewVieira replied:
I kind of did what you asked, but I'm not sure if you like my solution.
* (In most recent push)

CurtisGreen commented on src/targets/anvil.js:
```diff
@@ -27,46 +33,79 @@ export class Anvil extends ClickableObject {
         // See if bar is selected
         const selectedIndex = inv.curSelectedItemIndex;
-        if (selectedIndex < 0) {
+        const selectedItem = inv.inventory[selectedIndex];
```
Moving the conditional down means we might do `inv.inventory[-1]`, need to check for that before getting `selectedItem`

CurtisGreen replied:
Unless this no longer happens

AndrewVieira replied:
I fixed the issue, but I found another potential issue for future reference. If you are making an item and run out of material, the selected index from inventory is still valid, so the selected item returned by inventory is null. I changed the code to this:

`const selectedIndex = inv.curSelectedItemIndex;`
`if (selectedIndex < 0) {...}`

`const selectedItem = inv.inventory[selectedIndex];`
`if (!selectedItem || !this.validMaterials.has(selectedItem.name) {...}`

I should mention this error only happened after I removed async from the method.

AndrewVieira replied:
Async isn't related. I got the same error with the furnace when both tin and copper disappeared at the same time.

CurtisGreen commented on src/targets/anvil.js:
```diff
@@ -27,46 +33,79 @@ export class Anvil extends ClickableObject {
-                chat.writeText("The anvil can only be used with the bar selected");
-                break;
+    async selectRecipe(itemName) {
+        if (! await this.hasMaterials()) {
```
Just noticed this, why are the methods in this class and some of the other targets async? I don't see why it would be necessary since they don't await anything but maybe I'm misremembering something...

AndrewVieira replied:
I remember I was having issues without await here in this instance, but I don't remember the exact issue. I'll look at it again.

CurtisGreen replied:
Yeah, maybe because this.hasMaterial is a sync as well? If you don’t want to deal with it in this PR I can look into it later

CurtisGreen commented on src/ui/modals/smithing-modal-window.js:
```diff
@@ -0,0 +1,156 @@
+    getSpriteName(name) {
+        let result = name.toLowerCase();
+        result.toLowerCase();
+        result = result.replaceAll(" ", "-");
```
1) You have an extra `result.toLowerCase()` there.
2) I think it's cleaner to put: `return name.toLowerCase().replaceAll(" ", "-");`
3) Is this useful for future modals? Might want to put it in the parent class

AndrewVieira replied:
I put it in the parent modal class for now. I was thinking we might just make it a general function and place it in utilities.js.

CurtisGreen commented on src/ui/modals/smithing-modal-window.js:
```diff
@@ -0,0 +1,156 @@
+                this.elements.push(elementButton);
+            }
+
+            let elementIcon = this.scene.add
```
const

CurtisGreen commented on src/ui/modals/smithing-modal-window.js:
```diff
@@ -0,0 +1,156 @@
+
+            const textFont = levelFlag ? FONTS.SMITH_UNLOCKED : FONTS.SMITH_LOCKED;
+
+            let elementText = this.scene.add.text(positionX, positionY + this.iconHeight, item.item, textFont).setDepth(3);
```
const

CurtisGreen replied:
I wonder if you can chain a `setVisible(this.visible)` on the end so it can be all in one

CurtisGreen commented on src/ui/modals/smithing-modal-window.js:
```diff
@@ -0,0 +1,156 @@
+            const item = await getItemClass(element, this.scene.window);
+            const spriteName = await this.getSpriteName(item.name);
+
+            const barFlag = barSupply >= item.bars[0].count ? true : false;
```
Can be simplified to `const barFlag = barSupply >= item.bars[0].count;`

Same for one below this

CurtisGreen commented on src/ui/modals/smithing-modal-window.js:
```diff
@@ -0,0 +1,156 @@
+        let positionY = this.y + this.topOffset + this.verticalPadding;
+
+        for (let element of elements) {
+            if (element === "SKIP") {
```
Maybe use `null` instead so you can do `if (!element)`

CurtisGreen commented on src/ui/modals/smithing-modal-window.js:
```diff
@@ -0,0 +1,156 @@
+            const levelFlag = smithingLevel >= item.smithingLevel ? true : false;
+
+            if (barFlag && levelFlag) {
+                let elementButton = new Button(this.scene, positionX, positionY, this.buttonWidth, this.buttonHeight);
```
All of these objects can be changed to `const` from `let`

CurtisGreen commented on src/ui/modals/smithing-modal-window.js:
```diff
@@ -0,0 +1,156 @@
+            let elementText = this.scene.add.text(positionX, positionY + this.iconHeight, item.item, textFont).setDepth(3);
+            elementText.visible = this.visible;
+
+            let postfix = " bar";
```
`const postfix = (item.bars[0].count > 1) ? " bars" : " bar";`

CurtisGreen commented on src/ui/modals/smithing-modal-window.js:
```diff
@@ -0,0 +1,156 @@
+
+            if (rowIndex >= this.maxIconsPerRow) {
+                positionX = this.x + this.leftOffset + this.horizontalPadding;
+                positionY += this.iconHeight + this.verticalPadding + 24;
```
In general for loop creation logic like this I like to create these variables based off the index rather than adding to them.
Like instead of  `+=`  do:

```js
const yOffset = this.iconHeight + this.verticalPadding + 24;
const xOffset = ...
for (let [index, val] of array.entries()) {
    const positionY = index * yOffset;
    const positionX = index * xOffset;
    // ...
}
```

@EthanVieira Thoughts?

Another potential way of making this looping creation easier to read is mapping over the data and returning that into `elements` instead of creating elements and appending to it

EthanVieira replied:
I don't think I have a preference. Combining the terms into one offset does help readability though

CurtisGreen commented on src/ui/modals/smithing-modal-window.js:
```diff
@@ -0,0 +1,156 @@
+    async setChoices(bar, barSupply, smithingLevel) {
+        this.clearChoices();
+
+        let elements = [];
```
Not sure of a better name but having `elements` and `this.elements` is pretty confusing

CurtisGreen reviewed (changes requested):
Code looking good.

I just tested it and noticed two things
1) It's not very intuitive to select bars in inventory -> right click anvil -> choose type -> left click to forge. Should we prompt the window on what to forge every time or keep it as is?
2) Need to update `panels/skill-descriptions` to have the levels in which you can forge these new things

AndrewVieira commented:
One possible solution to reduce clicks is to auto-detect bars in the player's inventory if no bars are selected. The downside is that if there are multiple bar types, the type of bar auto-detected may not be the one the player wants to smith with. The player could still select the bar they want to smith with if they want.

Another solution could be to auto detect what bars the player has, then have multiple tabs for each bar type with the possible recipes shown for that bar when that tab is clicked. The downside to this is it would be a lot more effort to program and it differs much from the original Runescape interface.

CurtisGreen commented:
Yeah those could be good but they're a lot to add. I think we should keep it as-is for now and do an update later after some brainstorming

So just add the skill descriptions and I think it's good. Will have to see if Ethan finds anything else

EthanVieira commented on src/targets/anvil.js:
```diff
@@ -94,7 +134,9 @@ export class Anvil extends ClickableObject {
                 this.scene.stats.updateClickedTargetStat();
                 characterData.addSkillXp("smithing", item.xp);
-                this.scene.enemyKilled(item.questName);
+                if (this.scene.questExists(item.questName)) { 
```
Do we need this check here if we are also checking inside enemyKilled?

EthanVieira commented on src/test-smithing-data.js:
```diff
@@ -0,0 +1,121 @@
+    damageByAutoClick: 0,
+    numberOfAutoClickers: 0,
+    questPoints: 0,
+    inventory: [
```
Might want a large amount of ore as well if we think this smithing test data encompasses smelting as well. I'd also give a generous amount of gold just because why not :)

CurtisGreen replied:
💰

EthanVieira commented on src/ui/modals/smithing-modal-window.js:
```diff
@@ -0,0 +1,147 @@
+        }
+    }
+
+    getChoice() {
```
Don't think we need to override this method?

EthanVieira commented on src/test-smithing-data.js:
```diff
@@ -0,0 +1,121 @@
+export const testSmithingData = {
```
I think we should create a directory for data files

EthanVieira reviewed (changes requested):
Looking really good!

AndrewVieira commented:
I fixed the issues you guys mentioned. For the creation of icons I did it the way the Curtis mentioned, but it's a little more complicated as you have to account for the window position and limit of icons per row.

I think I don't want to add the remaining bronze items for the time being. I'm not sure if we will even use all the item types, and I kind of want to start working on a different features.

Let me know if you guys find any more issues. Hopefully we can merge this code soon.

EthanVieira commented on src/data/test-smithing-data.js:
```diff
@@ -16,8 +16,10 @@ export const testSmithingData = {
         { item: "Knife", count: 1 },
         { item: "Hammer", count: 1 },
-        { item: "Coin", count: 25 },
+        { item: "Coin", count: 69420 },
```
🥇

EthanVieira commented on src/ui/modals/smithing-modal-window.js:
```diff
@@ -62,48 +61,73 @@ export class SmithingModalWindow extends ModalWindow {
+            const positionY = offsetY + padY * parseInt(index / this.maxIconsPerRow);
 
             const item = await getItemClass(recipe, this.scene.window);
+            console.log(item.name);
```
Is this leftover?

AndrewVieira replied:
Oops I guess so, I'll wait for Curtis before updating

CurtisGreen replied:
Wait on me for what?

CurtisGreen replied:
I think we've done plenty on this PR already so anything else can probably be done later if needed

AndrewVieira replied:
I meant I was going to wait to see if you had any more changes in mind. I'll remove the console.log after I get off work today.

EthanVieira reviewed (commented):
Looks like maybe one leftover log statement, but besides that it all looks good to me!
